src/ctapipe/core/tool.py

```diff
@@ -425,26 +426,44 @@ def run(self, argv=None, raises=False):
 
                 self.start()
                 self.finish()
-                self.log.info(f"Finished: {self.name}")
+                self.log.info("Finished: %s", self.name)
                 Provenance().finish_activity(activity_name=self.name)
             except (ToolConfigurationError, TraitError) as err:
                 self.log.error("%s", err)
                 self.log.error("Use --help for more info")
                 exit_status = 2  # wrong cmd line parameter
+                Provenance().finish_activity(
+                    activity_name=self.name, status="error", exit_code=exit_status
+                )
                 if raises:
                     raise
             except KeyboardInterrupt:
                 self.log.warning("WAS INTERRUPTED BY CTRL-C")
+                exit_status = 130  # Script terminated by Control-C
                 Provenance().finish_activity(
-                    activity_name=self.name, status="interrupted"
+                    activity_name=self.name, status="interrupted", exit_code=exit_status
                 )
-                exit_status = 130  # Script terminated by Control-C
             except Exception as err:
-                self.log.exception(f"Caught unexpected exception: {err}")
-                Provenance().finish_activity(activity_name=self.name, status="error")
+                self.log.exception("Caught unexpected exception: %s", err)
                 exit_status = 1  # any other error
+                Provenance().finish_activity(
+                    activity_name=self.name, status="error", exit_code=exit_status
+                )
                 if raises:
                     raise
+            except SystemExit as err:
```

Comment:
Do we need to check for 0 as a special case of success here?

Reply:
hm... why would someone call a `sys.exit(0)` inside a tool? a normal termination is done later... I think such behavior should be banned.

Reply:
There could be an early exit success. But that would also be possible with just `return`

Reply:
yes, exactly. And I believe, this shall be the proper way to go.